src/antares/service/local_services/binding_constraint_local.py

```diff
@@ -103,20 +95,76 @@ def _check_if_empty_ts(time_step: BindingConstraintFrequency, time_series: Optio
-            for idx, idx_constraint in enumerate(self.binding_constraints.values())
-        }
-        self.ini_file.ini_dict = binding_constraints_ini_content
+    def _write_binding_constraint_ini(
+        self,
+        properties: BindingConstraintProperties,
+        constraint_name: str,
+        constraint_id: str,
+        terms: Optional[list[ConstraintTerm]] = None,
+    ) -> None:
+        """
+        Write a single binding constraint to the INI file, reconstructing a full BindingConstraintPropertiesLocal instance.
+
+        Args:
+            properties (BindingConstraintProperties): Basic properties of the binding constraint.
+            constraint_name (str): The name of the constraint.
+            constraint_id (str): The ID of the constraint.
+            terms (dict[str, ConstraintTerm], optional): Terms applying to the binding constraint. Defaults to None.
+        Raises:
+        BindingConstraintCreationError: If a binding constraint with the same name already exists in the INI file.
+        """
+
+        current_ini_content = self.ini_file.ini_dict or {}
+
+        if constraint_name in current_ini_content:
+            if terms in [None, {}]:
+                raise BindingConstraintCreationError(
+                    constraint_name=constraint_name,
+                    message=f"A binding constraint with the name {constraint_name} already exists with terms.",
+                )
+
+        terms_dict = {term.id: term for term in terms} if terms else {}
+
+        full_properties = BindingConstraintPropertiesLocal(
+            constraint_name=constraint_name, constraint_id=constraint_id, terms=terms_dict, **properties.model_dump()
+        )
+
+        current_ini_content = self.ini_file.ini_dict or {}
+
+        current_ini_content[full_properties.constraint_name] = full_properties.list_ini_fields
+
+        self.ini_file.ini_dict = current_ini_content
+
         self.ini_file.write_ini_file()
 
     def add_constraint_terms(self, constraint: BindingConstraint, terms: list[ConstraintTerm]) -> list[ConstraintTerm]:
-        new_terms = constraint.local_properties.terms | {
-            term.id: term for term in terms if term.id not in constraint.get_terms()
+        """
+        Add terms to a binding constraint and update the INI file.
+
+        Args:
+            constraint (BindingConstraint): The binding constraint to update.
+            terms (list[ConstraintTerm]): A list of new terms to add.
+
+        Returns:
+            list[ConstraintTerm]: The updated list of terms.
+        """
+
+        new_terms = {
+            **constraint.local_properties.terms,  # Existing terms
+            **{term.id: term for term in terms if term.id not in constraint.get_terms()},  # New terms
         }
+
         constraint.local_properties.terms = new_terms
-        self._write_binding_constraint_ini()
+
+        list(new_terms.values())
+
+        self._write_binding_constraint_ini(
```

Comment:
Minor perf improvement: you could store list(new_terms.values()) inside a variable to avoid doing this twice (lines 165 and 168)

Reply:
You can change line 208 to return the introduced variables terms_values

Reply:
yes! done